File: src/api/restriction_diagnostics.py
"""Passive risk-lock / config restriction fields for dashboard diagnostics."""

from __future__ import annotations

from typing import Any


def _live_config_restrictions() -> dict[str, Any]:
    try:
        from system.config_loader import get_config

        cfg = get_config()
        max_open = int(cfg.max_open_positions)
        rotation_on = bool(cfg.get("enforce_top3_rotation_filter", True))
    except Exception:
        max_open = 0
        rotation_on = True
    return {
        "max_open_positions": max_open,
        "enforce_top3_rotation_filter": rotation_on,
    }


def enrich_restrictions_payload(payload: dict[str, Any]) -> dict[str, Any]:
    """
    Inject ``config`` and ``system_state.hydration.max_open_positions`` for UI standby banner.
    """
    out = dict(payload)
    restrictions = _live_config_restrictions()
    out["config"] = dict(restrictions)

    sys_snap = dict(out.get("system_state") or {})
    hydration = dict(sys_snap.get("hydration") or {})
    hydration["max_open_positions"] = restrictions["max_open_positions"]
    sys_snap["hydration"] = hydration
    out["system_state"] = sys_snap
    return out

File: dashboard/src/utils/restrictionDiagnostics.js
/**
 * Passive config restriction diagnostics — /api/health & /api/startup/status.
 */

const STANDBY_MESSAGE =
  "⚠️ SYSTEM STANDBY: Sizing Restricted (POS 0/0) / Rotation Filter Engaged";

export function resolveRestrictionDiagnostics(payload) {
  if (!payload || typeof payload !== "object") {
    return {
      active: false,
      sizingRestricted: false,
      rotationFilterEngaged: false,
      maxOpenPositions: null,
      message: "",
    };
  }

  const hydration = payload.system_state?.hydration || {};
  const config = payload.config || {};
  const maxOpen =
    hydration.max_open_positions ??
    config.max_open_positions ??
    // /state has top-level max_open_positions even when /api/health enrichment is not warmed.
    payload.max_open_positions ??
    null;
  const rotationFilter =
    config.enforce_top3_rotation_filter ??
    payload.system_state?.config?.enforce_top3_rotation_filter;

  // Treat "missing" as unknown (do not lock UI into POS 0/0 by default).
  const sizingRestricted = maxOpen !== null && Number(maxOpen) === 0;
  const rotationFilterEngaged = rotationFilter === true;
  const active = sizingRestricted || rotationFilterEngaged;

  return {
    active,
    sizingRestricted,
    rotationFilterEngaged,
    maxOpenPositions: maxOpen,
    message: active ? STANDBY_MESSAGE : "",
  };
}

export function isSystemStandbyRestricted(payload) {
  return resolveRestrictionDiagnostics(payload).active;
}

export { STANDBY_MESSAGE };
